src/hotspot/share/runtime/handshake.hpp

```diff
@@ -61,33 +63,49 @@ class Handshake : public AllStatic {
   // Execution of handshake operation
   static void execute(HandshakeClosure*       hs_cl);
   static void execute(HandshakeClosure*       hs_cl, JavaThread* target);
-  static void execute(AsyncHandshakeClosure* hs_cl, JavaThread* target);
+  static void execute(AsyncHandshakeClosure*  hs_cl, JavaThread* target);
 };
 
+class JvmtiRawMonitor;
+
 // The HandshakeState keeps track of an ongoing handshake for this JavaThread.
 // VMThread/Handshaker and JavaThread are serialized with _lock making sure the
 // operation is only done by either VMThread/Handshaker on behalf of the
 // JavaThread or by the target JavaThread itself.
 class HandshakeState {
+  friend JvmtiRawMonitor;
+  friend ThreadSuspensionHandshake;
+  friend SuspendThreadHandshake;
+  friend JavaThread;
   // This a back reference to the JavaThread,
   // the target for all operation in the queue.
   JavaThread* _handshakee;
   // The queue containing handshake operations to be performed on _handshakee.
   FilterQueue<HandshakeOperation*> _queue;
-  // Provides mutual exclusion to this state and queue.
-  Mutex   _lock;
+  // Provides mutual exclusion to this state and queue. Also used for
+  // JavaThread suspend/resume operations.
+  Monitor _lock;
   // Set to the thread executing the handshake operation.
   Thread* _active_handshaker;
 
   bool claim_handshake();
   bool possibly_can_process_handshake();
   bool can_process_handshake();
-  void process_self_inner();
+
+  // Returns false if the JavaThread finished all its handshake operations.
+  // If the method returns true there is still potential work to be done,
+  // but we need to check for a safepoint before.
+  // (this is due to a suspension handshake which put the JavaThread in blocked
+  // state so a safepoint may be in-progress)
```

Comment:
nit: s/(this/(This/
nit: s/in-progress)/in-progress.)/

Reply:
Fixed